irods/test/pam_interactive_test.py

```diff
@@ -74,5 +76,85 @@ def test_failed_login_incorrect_password(self):
+            ("replace_resp_fallback", {"msg": {"patch": [{"op": "replace", "path": "/username"}]}, "pstate": {"username": "rods"}, "resp": "alice"}, {"username": "alice"}, True),
+            ("nested_add_operation", {'msg': {'patch': [{'op': 'add', 'path': '/user/name', 'value': 'alice'}]}, 'pstate': {'user': {}}}, {'user': {'name': 'alice'}}, True),
+            ("resp_fallback_empty", {'msg': {'patch': [{'op': 'add', 'path': '/username'}]}, 'pstate': {}}, {'username': ''}, True),
+            ("no_patch_ops", {"msg": {}, "pstate": {"username": "alice"}, "pdirty": False}, {"username": "alice"}, False)
+        ]
+
+        for name, request, expected_pstate, expected_pdirty in test_cases:
+            with self.subTest(name=name):
+                self.auth_client._patch_state(request)
+                self.assertEqual(request["pstate"], expected_pstate)
+                self.assertEqual(request["pdirty"], expected_pdirty)
+
+    def test_retrieve_entry(self):
+        test_cases = [
+            ("surface_value", {"msg": {"retrieve": "/user"}, "pstate": {"user": "alice"}}, True, "alice"),
+            ("nested_value", {"msg": {"retrieve": "/user/password"}, "pstate": {"user": {"password": "rods"}}}, True, "rods"),
+            ("empty_value", {"msg": {"retrieve": "/user"}, "pstate": {"user": ""}}, True, ""),
+            ("path_does_not_exist", {"msg": {"retrieve": "/missing"}, "pstate": {"user": "alice"}}, True, ""),
+            ("non_string_value", {'msg': {'retrieve': '/user/id'}, 'pstate': {'user': {'id': 456}}}, True, '456'),
+            ("no_retrieve_key", {"msg": {}, "pstate": {"user": "alice"}}, False, None)
+        ]
+
+        for name, request, expected_result, expected_resp in test_cases:
+            with self.subTest(name=name):
+                self.assertEqual(self.auth_client._retrieve_entry(request), expected_result)
+                if expected_result:
+                    self.assertEqual(request["resp"], expected_resp)
+
+    def test_get_input(self):
+        test_cases = [
+            ("non_password_input", False, 'sys.stdin.readline', "rods\n", {"msg": {"prompt": "Prompt:"}}, "rods"),
+            ("non_password_default", False, 'sys.stdin.readline', "\n", {"msg": {"prompt": "Prompt:", "default_path": "/password"}, "pstate": {"password": "rods"}}, "rods"),
+            ("password_input", True, 'getpass.getpass', "rods", {"msg": {"prompt": "Password:"}}, "rods"),
+            ("password_default", True, 'getpass.getpass', "", {"msg": {"prompt": "Password:", "default_path": "/password"}, "pstate": {"password": "rods"}}, "rods")
+        ]
+
+        for name, is_password, mock_target, user_input, request, expected_resp in test_cases:
+            with self.subTest(name=name), patch('irods.auth.pam_interactive._auth_api_request', return_value={"result": "ok"}), \
+            patch(mock_target, return_value=user_input), patch('sys.stderr.write'), patch.object(self.auth_client, '_patch_state') as mock_patch:
```

Comment:
This is difficult to read, so I think it would be good to think about how we might improve this. Maybe it would make sense to use the decorator form of `unittest.mock.patch` for the patches which do not require local objects (e.g. `patch('sys.stderr.write')`). Thoughts?

Also, `mock_patch` appears to be unused.

Reply:
D'oh, I see it now. Thanks